### PR TITLE
[AMORO-2276]: UnifiiedCatalog for Spark Engine

### LINKED ISSUE #2276: [Subtask][Spark]: Support UnifiedCatalog in Spark Engine.
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### Description

_No response_

### Parent issue

#1061

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by baiyangtx**
## Why are the changes needed? 

Close #2276.

## Brief change log 

- Add UnifiedSparkCatalog under spark common module
- Extract MixedSparkCatalogBase and MixedSparkSessionCatalogBase to spark common module
- Refactor spark unit test framework to adapt unifed catalog tests and mixed format tests.

## How was this patch tested?

- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: hive/src/test/java/com/netease/arctic/hive/catalog/HiveCatalogTestHelper.java [hunk 1]
```diff
@@ -72,9 +67,8 @@ public TableFormat tableFormat() {
   @Override
   public CatalogMeta buildCatalogMeta(String baseDir) {
     Map<String, String> properties = Maps.newHashMap();
-    if (TableFormat.MIXED_ICEBERG == tableFormat) {
-      properties.put(CatalogProperties.URI, hiveConf.get(HiveConf.ConfVars.METASTOREURIS.varname));
-    }
+    properties.put(CatalogProperties.URI, hiveConf.get(HiveConf.ConfVars.METASTOREURIS.varname));
```
Comment (YesOrNo828) [inline hive/src/test/java/com/netease/arctic/hive/catalog/HiveCatalogTestHelper.java:70]: Should split these codes first? I think the Flink engine also need it.

**Review by wangtaohz — COMMENTED** [review]
I left some comments, please take a look. @baiyangtx

File: core/src/main/java/com/netease/arctic/CommonUnifiedCatalog.java [hunk 2]
```diff
@@ -41,16 +41,29 @@ public class CommonUnifiedCatalog implements UnifiedCatalog {
...
     initializeFormatCatalogs();
   }
 
+  @Override
+  public String metastoreType() {
+    return meta.getCatalogType();
+  }
+
+  @Override
+  public TableMetaStore authenticationContext() {
```
Comment (YesOrNo828) [inline core/src/main/java/com/netease/arctic/CommonUnifiedCatalog.java:63]: When I try to complete this #2344 Flink Unified Catalog(Iceberg format table), I find out that this API is aslo needed.

File: mixed/spark/v3.3/spark/pom.xml [hunk 3]
```diff
@@ -259,6 +259,18 @@
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.paimon</groupId>
+            <artifactId>paimon-spark-3.1</artifactId>
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/v3.3/spark/pom.xml:264]: The engine version is wrong. I will fix that

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/MixedFormatSparkUtil.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * Check a Spark Table is a mixed-format table.
+   *
+   * @param table the spark table loaded from session catalog
+   * @return is it a mixed format table.
+   */
+  public static boolean isMixedFormatTable(Table table) {
+    boolean isMixedHive = isMixedHiveTable(table);
+    boolean isMixedIceberg = isMixedIcebergTable(table);
+    return isMixedHive || isMixedIceberg;
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/MixedFormatSparkUtil.java:38]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/MixedFormatSparkUtil.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  /**
+   * Check a Spark Table is a Mixed Iceberg table
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/MixedFormatSparkUtil.java:42]: fixed

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/paimon/PaimonSparkFormat.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+
+package com.netease.arctic.spark.paimon;
+
+import com.netease.arctic.ams.api.TableFormat;
+import com.netease.arctic.spark.SparkTableFormat;
+import org.apache.spark.sql.connector.catalog.Table;
+
+public class PaimonSparkFormat implements SparkTableFormat {
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/paimon/PaimonSparkFormat.java:25]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/iceberg/IcebergSparkFormat.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  @Override
+  public TableFormat format() {
+    return TableFormat.ICEBERG;
+  }
+
+  @Override
+  public boolean isSubTableName(String tableName) {
+    return MetadataTableType.from(tableName) != null
+        || AT_TIMESTAMP.matcher(tableName).matches()
+        || SNAPSHOT_ID.matcher(tableName).matches();
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/iceberg/IcebergSparkFormat.java:44]: Those logical is copied from iceberg projects.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SupportAuthentication.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+
+package com.netease.arctic.spark;
+
+import com.netease.arctic.table.TableMetaStore;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+
+/** Mixed-in interface to indicate that a spark catalog support Amoro authentication context. */
+public interface SupportAuthentication extends TableCatalog {
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SupportAuthentication.java:25]: This is a mixed-in interface style. to limit this interface is only used for catalog.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+  private TableCatalog initializeTableCatalog(TableFormat format) {
+    String catalogOptions = format.name().toLowerCase().replace("_", "-") + ".spark-catalog-impl";
+    String impl = unifiedCatalog.properties().get(catalogOptions);
+    if (StringUtils.isBlank(impl)) {
+      impl = defaultTableCatalogImplMap.get(format);
+    }
+    if (StringUtils.isBlank(impl)) {
+      throw new IllegalStateException(
+          "failed to initialize spark TableCatalog for format:" + format.name());
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java:287]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    Preconditions.checkNotNull(uri, "lack required option: %s", SparkUnifiedCatalogProperties.URI);
+
+    ArcticThriftUrl catalogUri = ArcticThriftUrl.parse(uri, Constants.THRIFT_TABLE_SERVICE_NAME);
+    String registerCatalogName = catalogUri.catalogName();
+
+    if (StringUtils.isBlank(registerCatalogName)) {
+      registerCatalogName = name;
+      if (CatalogManager.SESSION_CATALOG_NAME().equalsIgnoreCase(registerCatalogName)) {
+        LOG.warn(
+            "catalog name is not exists in catalog uri, using spark catalog as register catalog name, but "
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java:89]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/util/SparkUtil.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.connector.catalog.CatalogManager;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+
+import java.util.List;
+
+public class SparkUtil {
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/util/SparkUtil.java:30]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalogProperties.java [hunk 12]
```diff
@@ -18,12 +18,8 @@
 
 package com.netease.arctic.spark;
 
-import java.util.Locale;
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalogProperties.java:21]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/MixedSessionCatalogBase.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      String name, CaseInsensitiveStringMap options);
+
+  @Override
+  protected boolean isManagedTable(Table table) {
+    return MixedFormatSparkUtil.isMixedFormatTable(table);
+  }
+
+  @Override
+  protected boolean isManagedProvider(String provider) {
+    return "arctic".equalsIgnoreCase(provider);
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/MixedSessionCatalogBase.java:47]: fixed.

File: core/src/main/java/com/netease/arctic/TableIDWithFormat.java [hunk 14]
```diff
@@ -42,4 +42,12 @@ public TableIdentifier getIdentifier() {
   public TableFormat getTableFormat() {
     return tableFormat;
   }
+
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply core/src/main/java/com/netease/arctic/TableIDWithFormat.java:45]: I'm quite skeptical whether the "catalog" in the class TableIdentifier is necessary. It seems that there is nowhere to use this value, and it brings in a lot of unnecessary checks.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/paimon/PaimonSparkFormat.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+  final String KEY_STORAGE_HANDLER = "storage_handler";
+  final String PAIMON_STORAGE_HANDLER = "org.apache.paimon.hive.PaimonStorageHandler";
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/paimon/PaimonSparkFormat.java:27]: fixed

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    Table table = getSessionCatalog().loadTable(ident);
+    if (isManagedTable(table)) {
+      return getTargetCatalog().loadTable(ident);
+    }
+    return table;
+  }
+
+  protected boolean isManagedSubTable(Identifier ident) {
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:165]: fixed

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+  }
+
+  /**
+   * build mixed-format catalog instance.
+   *
+   * @param name spark catalog name
+   * @param options catalog initialize options
+   * @return mixed format spark catalog.
+   */
+  protected abstract TableCatalog buildTargetCatalog(String name, CaseInsensitiveStringMap options);
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:100]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.catalog.TableChange;
+import org.apache.spark.sql.connector.expressions.Transform;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+import java.util.Map;
+
+/** Base class of spark session catalog. */
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:40]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkTableFormat.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+  /** Table format */
+  TableFormat format();
+
+  /** Check if the give table name is a valid inspecting table name. */
+  default boolean isSubTableName(String tableName) {
+    return false;
+  }
+
+  /** Check the table loaded by spark session catalog is match the table format. */
+  boolean isSessionTable(Table table);
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkTableFormat.java:36]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/paimon/PaimonSparkFormat.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wangtaohz) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/paimon/PaimonSparkFormat.java:25]: Still missing.

File: mixed/hive/src/test/java/com/netease/arctic/hive/catalog/HiveCatalogTestHelper.java [hunk 20]
```diff
@@ -45,16 +44,11 @@ public class HiveCatalogTestHelper implements CatalogTestHelper {
   private final Configuration hiveConf;
 
   public static CatalogTestHelper build(Configuration hiveConf, TableFormat... formats) {
-    Preconditions.checkArgument(formats.length == 1, "Cannot support multiple table formats");
+    // Preconditions.checkArgument(formats.length == 1, "Cannot support multiple table formats");
```
[the remark this answers is not in the document]
Comment (wangtaohz) [reply mixed/hive/src/test/java/com/netease/arctic/hive/catalog/HiveCatalogTestHelper.java]: This line has not been removed yet. 
``` java
// Preconditions.checkArgument(formats.length == 1, "Cannot support multiple table formats");
```

File: core/src/main/java/com/netease/arctic/TableIDWithFormat.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (baiyangtx) [reply core/src/main/java/com/netease/arctic/TableIDWithFormat.java:45]: This is a requirement for ServerCatalog. In fact, the catalog field is only needed in ServerTableIdentifier. It is not required elsewhere.

File: mixed/spark/common/pom.xml [hunk 21]
```diff
@@ -215,43 +215,16 @@
 
     </dependencies>
 
+    <build>
```
Comment (zhoujinsong) [inline mixed/spark/common/pom.xml:218]: I haven't seen the new Scala code. Why do we need this plugin?

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+    if (isManagedTable(table)) {
+      return getTargetCatalog().alterTable(ident, changes);
+    } else {
+      return getSessionCatalog().alterTable(ident, changes);
+    }
+  }
+
+  @Override
+  public boolean dropTable(Identifier ident) {
+    // no need to check table existence to determine which catalog to use. if a table doesn't exist
```
Comment (zhoujinsong) [inline mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:197]: The comments here seem to have been broken due to automatic formatting.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+        return getSessionCatalog().dropTable(ident);
+      }
+    } catch (NoSuchTableException e) {
+      return getSessionCatalog().dropTable(ident);
+    }
+  }
+
+  @Override
+  public boolean purgeTable(Identifier ident) throws UnsupportedOperationException {
+    // no need to check table existence to determine which catalog to use. if a table doesn't exist
```
Comment (zhoujinsong) [inline mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:214]: The comments here seem to have been broken due to automatic formatting. You may want to set the IDE's maximum line length to 100 to match the configured auto-formatting.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+      }
+    } catch (NoSuchTableException e) {
+      return getSessionCatalog().purgeTable(ident);
+    }
+  }
+
+  @Override
+  public void renameTable(Identifier from, Identifier to)
+      throws NoSuchTableException, TableAlreadyExistsException {
+    // rename is not supported by HadoopCatalog. to avoid UnsupportedOperationException for session
```
Comment (zhoujinsong) [inline mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:232]: The comments here seem to have been broken due to automatic formatting.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java [hunk 25]
```diff
@@ -0,0 +1,338 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+public class SparkUnifiedCatalog implements TableCatalog, SupportsNamespaces, ProcedureCatalog {
```
Comment (zhoujinsong) [inline mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java:60]: Please add some Java docs for the new class.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java [hunk 26]
```diff
@@ -0,0 +1,338 @@
...
+    return name;
+  }
+
+  private String namespaceToDatabase(String[] namespace) {
+    Preconditions.checkArgument(namespace.length == 1, "only support namespace with 1 level.");
+    return namespace[0];
+  }
+
+  @Override
+  public String[][] listNamespaces() throws NoSuchNamespaceException {
```
Comment (zhoujinsong) [inline mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java:116]: We should remove unnecessary throw declarations.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java [hunk 27]
```diff
@@ -0,0 +1,338 @@
...
+    return unifiedCatalog.listDatabases().stream()
+        .map(d -> new String[] {d})
+        .toArray(String[][]::new);
+  }
+
+  @Override
+  public String[][] listNamespaces(String[] namespace) throws NoSuchNamespaceException {
```
Comment (zhoujinsong) [inline mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java:123]: We should remove unnecessary throw declarations.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java (hunk 26 above)
Comment (zhoujinsong) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java:116]: I found there are other same issues in this class, can you fix them all?

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/ArcticDataSource.java [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package com.netease.arctic.spark.mixed;
+
+/** Arctic DataSource register */
+public class ArcticDataSource extends MixedSourceBase {
+
+  @Override
+  public String shortName() {
+    return "arctic";
```
Comment (zhoujinsong) [inline mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/ArcticDataSource.java:26]: Is `arctic` still a proper name for the source? We have changed the project name to amoro and the project name is no longer related to a specific source.

**Review by zhoujinsong — COMMENTED** [review]
@baiyangtx I left some comments. Please take a look.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java (hunk 26 above)
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SparkUnifiedCatalog.java]: fixed

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/ArcticDataSource.java (hunk 28 above)
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/mixed/ArcticDataSource.java:31]: fixed

File: mixed/spark/common/pom.xml (hunk 21 above)
Comment (baiyangtx) [reply mixed/spark/common/pom.xml:218]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java (hunk 22 above)
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:197]: fixed.

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java (hunk 23 above)
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:214]: fixed

File: mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java (hunk 24 above)
Comment (baiyangtx) [reply mixed/spark/common/src/main/java/com/netease/arctic/spark/SessionCatalogBase.java:232]: fixed

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**